File: src/rag_production_level/Docs/rag_document_corpus/code/csharp/DocumentService.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Nexus.Documents;

public sealed class DocumentService
{
    private readonly IDocumentRepository _repository;

    public DocumentService(IDocumentRepository repository)
    {
        _repository = repository;
    }

    public async Task<Document> CreateAsync(
        Document document,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(document.Title))
            throw new ArgumentException("Title is required.");

        var existing = await _repository.FindByHashAsync(
            document.TenantId,
            document.ContentHash,
            cancellationToken);

        if (existing is not null)
            return existing;

        return await _repository.InsertAsync(document, cancellationToken);
    }
}

public interface IDocumentRepository
{
    Task<Document?> FindByHashAsync(string tenantId, string hash, CancellationToken ct);
    Task<Document> InsertAsync(Document document, CancellationToken ct);
}

public sealed record Document(
    string Id,
    string TenantId,
    string Title,
    string ContentHash);

File: src/rag_production_level/Docs/rag_document_corpus/code/python/document_service.py
from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class Document:
    document_id: str
    tenant_id: str
    title: str
    content_hash: str


class DocumentRepository(Protocol):
    def find_by_hash(self, tenant_id: str, content_hash: str) -> Document | None: ...
    def insert(self, document: Document) -> Document: ...


class DocumentService:
    def __init__(self, repository: DocumentRepository) -> None:
        self.repository = repository

    def create(self, document: Document) -> Document:
        if not document.title.strip():
            raise ValueError("title is required")

        existing = self.repository.find_by_hash(
            document.tenant_id, document.content_hash
        )
        if existing:
            return existing

        return self.repository.insert(document)
